nixos/modules/services/misc/mautrix-telegram.nix

```diff
@@ -0,0 +1,155 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  dataDir = "/var/lib/mautrix-telegram";
+in {
+  options = {
+    services.mautrix-telegram = {
+      enable = mkEnableOption "Mautrix-Telegram, a Matrix-Telegram hybrid puppeting/relaybot bridge";
+
+      config = mkOption rec {
+        type = types.attrs;
+        apply = recursiveUpdate default;
```

Comment:
No need to put this in `apply`, you can instead set the defaults by doing
```
{
  config.services.mautrix-telegram.config = {
    appservice = {
      database = "sqlite:///${dataDir}/mautrix-telegram.db";
      hostname = mkDefault "0.0.0.0";
      # ...
    };
  };
}
```

This makes merging values work correctly and allows a lot more control over what should be defaulted and how. The downside is that such defaults aren't visible in the docs, but this should be fixed in the module system eventually.

See also [this section](https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md#defaults) which talks about when to use `mkDefault` or not, from [my RFC](https://github.com/NixOS/rfcs/pull/42) about such NixOS options. That section also recommends making the default as minimal as possible to get a working service, which I recommend you do here too (if you haven't already).

Reply:
I changed the module accordingly.

It seems however that `mkDefault` doesn't play well with `toJSON`:

```nix
config.services.mautrix-telegram.config.logging.root = {
        version = 1;

        formatters.precise.format = "[%(levelname)s@%(name)s] %(message)s";

        handlers.console = {
          class = "logging.StreamHandler";
          formatter = "precise";
        };

        loggers = {
          mau.level = mkDefault "INFO";
          telethon.level = mkDefault "INFO";
          aiohttp.level = mkDefault "INFO";
        };

        root = {
          level = mkDefault "INFO";
          handlers = [ "console" ];
        };
};
```

`toJSON` generates:
```yaml
logging: {formatters: {precise: {format: '[%(levelname)s@%(name)s] %(message)s'}},
    handlers: {console: {class: logging.StreamHandler, formatter: precise}}, loggers: {
        aiohttp: {level: {_type: override, content: INFO, priority: 1000}}, mau: {
            level: {_type: override, content: INFO, priority: 1000}}, telethon: {
            level: {_type: override, content: INFO, priority: 1000}}}, root: {handlers: [
            console], level: {_type: override, content: INFO, priority: 1000}}, version: 1}
```

(note the unwanted `_type`, `content` and `priority` fields).

Reply:
It also appear that the attribute set isn't recursively merged.
Nested definitions outside of the module override the whole parent attribute set.

For instance, setting `services.mautrix-telegram.logging.root.level = "DEBUG"` causes all the default attributes defined in `config.services.mautrix-telegram.config.logging` inside the module to be forgotten.

Reply:
`types.attrs` doesn't merge recursively:
* "Telegraf extraConfig is not recursively merged": https://github.com/NixOS/nixpkgs/issues/32217
* https://github.com/NixOS/nixpkgs/blob/60e8fcf0e56ce881294b06bd1f40069016308162/lib/types.nix#L226
* https://github.com/NixOS/nixpkgs/blob/6cf583cf2f8d3cb89fa8a4d6edd86f4236862ea6/lib/trivial.nix#L79

Reply:
Yeah so you'd need to use a fully specified type as described in the RFC for this to work properly, forgot about that. So for now with `types.attrs` just setting all values without a mkDefault might be best. Unfortunately when users set certain values, all the defaults under that attr get cleared, but that's been the case for all options that use `types.attrs` previously too, so I'd say we just use that anyways for now.

I should note in the RFC that when it's merged we should go over this module to specify the type properly and set the `mkDefault`'s.